### PR TITLE
Learn by example intro + first two posts

### PR DESCRIPTION
**Pull request by JorgeCastilloPrz**
The overall goal for the "Learn by example" site is going over patterns to compose a program gradually based on the example, so we can have a single series of blogposts targeting the latest Arrow version that anyone can use as a cheat sheet to write a pure functional application. Each post will add more features to the program on top of the previous one, so people goes sequentially from one to the other and we can have "pattern based guides" closer to what real world examples would be. The aim is to close the gap for newcomers and get closer to their day to day needs.

**This site will replace the patterns section from core docs we have now.**

In this case I'm covering the first two:

1. Writing an initial program
  * [x] Tech review
  * [x] Lang review
2. Handling errors
  * [x] Tech review
  * [x] Lang review

Following ones will be

3. Modelling data
4. Safe background computation (IO)
5. Independent computations
6. Dependent (sequential) computations
7. Concurrency
8. Traversing container types
9. Making our program poylmorphic
10. Injecting dependencies
Writing tests for our program

The final list is prone to change and we'll be figuring out, but that's an initial draft of the available sections. It will also be a second series for common integrations that will likely take this app and integrate it with Arrow.

### CODE REVIEW

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  println("UserId: $userId")
+  val user = InMemoryUserDatabase.findUser(userId)
+  println(user)
+  //sampleEnd
+}
+``` 
+
+As you can see, we could register and find a user in the database. But this program has a few issues.
+
+We are not handling any potential errors, because we assumed it's an in memory implementation that will not fail. But what If it was a real database? **It could potentially throw, making our function calls non deterministic**. Also, what happens when the user we are trying to find is not there? If our backend was a Resful API, we would probably want to handle it by returning a 404 not found response to the client.
```
Comment (raulraja) [inline docs/docs/learn-by-example/1-writing-an-initial-program/README.md]: perhaps introduce that the call could be async and the current implementation forces to block

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+We will cover a set of usual concepts like asynchrony, concurrency, controlling effects, parallel and sequential computation, thread switching, domain data and error modeling, resource safety, dependency injection, testing and much more.
+
+## 1. Writing the initial program
+
+Along this series of posts we will gradually build a backend for a Rock Band social network 🎸🤘
+
+<img src="/img/learn-by-example/band_playing.gif" alt="Rock band playing" width="800"/>
+
+Let's start by writing an early simple version of our program to start from there. Let's image a `UserDatabase` for the social network so new users can register:
```
Comment (raulraja) [inline docs/docs/learn-by-example/1-writing-an-initial-program/README.md]:
```suggestion
Let's start by writing an early simple version of our program to start from there. Let's model a `UserDatabase` for the social network so new users can register:
```

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+Ultimately we need to account for errors in our domain and handle them properly to make our program resilient, the same way a rock band would need to keep playing if something unexpected happened.
+
+<img src="/img/learn-by-example/band_error_keep_playing.gif" alt="Rock band playing" width="800"/>
+
+On top of that, it's a database, so by definition it represents **a mutable state**. Even if our implementation is in-memory, we are not reflecting this mutability with the public function types in the contract.
+
+Every time we create a user, the database internal state will change. If we keep in mind that the `createUser` function could be **called from multiple places** in our program, this encoding would make it hard for us the developers to track down what the state of our program is at any point in time.
+
+Overall, we are introducing ambiguity in our program provoked by a **"side effect"**, which makes the program non deterministic. That blocks our ability to apply *local reasoning* over pieces of logic relying on it. And if we can't reason about those atomic pieces, we will not be able to reason over bigger logics relying on them, and ultimately over our program as a whole.
```
Comment (raulraja) [inline docs/docs/learn-by-example/1-writing-an-initial-program/README.md]:
```suggestion
Overall, we are introducing ambiguity in our program provoked by a [**"side effect"**](https://en.wikipedia.org/wiki/Side_effect_(computer_science)), which makes the program [non-deterministic](https://en.wikipedia.org/wiki/Nondeterministic_algorithm). That blocks our ability to apply *local reasoning* over pieces of logic relying on it. And if we can't reason about those atomic pieces, we will not be able to reason over bigger programs relying on them, and ultimately over our system as a whole.
```

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+Finally, we have a second side effect imposed by the `generateId()` function. That function returns a different value every time we call it, so it is impure by definition.
```
Comment (raulraja) [inline docs/docs/learn-by-example/1-writing-an-initial-program/README.md]:
```suggestion
Finally, we have a second side effect imposed by the `generateId()` function. That function returns a different value every time we call it, so it is impure and non-deterministic by definition.
```

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+Let's start by writing an early simple version of our program to start from there. Let's image a `UserDatabase` for the social network so new users can register:
+
+```kotlin
+interface UserDatabase {
+  fun createUser(name: String): UserId
+  fun findUser(userId: UserId): User?
+}
+```
+
+If our backend was a **Resful API**, we would expect it to offer an endpoint to create new users that would rely on this database.
```
Comment (rachelcarmena) [inline docs/docs/learn-by-example/1-writing-an-initial-program/README.md]:
```suggestion
If our backend was a **RESTful API**, we would expect it to offer an endpoint to create new users that would rely on this database.
```

**Review by raulraja — APPROVED** [review]
Looks great @JorgeCastilloPrz ! Still hard to see as a whole until we go over the next examples because I feel IO has to show up here soon to do anything useful or we would make the same mistake of introducing IO last when we are actually expecting that for 99% of cases they need suspend or IO

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md (hunk 1 above)
Comment (JorgeCastilloPrz) [reply docs/docs/learn-by-example/1-writing-an-initial-program/README.md]: Right, added now 👍

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md (hunk 2 above)
Comment (JorgeCastilloPrz) [reply docs/docs/learn-by-example/1-writing-an-initial-program/README.md]: Not sure what was on my mind there, maybe "imagine" :D

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md (hunk 3 above)
Comment (JorgeCastilloPrz) [reply docs/docs/learn-by-example/1-writing-an-initial-program/README.md]: Awesome, thanks 👍

**Comment by JorgeCastilloPrz** [comment]
Thanks @raulraja. Went over all your comments. Yep right now I've got a further post for deferring execution on the index, but might definitely push it up a bit so it's closer to beginning. Many of these posts might change ordering and be refactored the more I complete content, since the overall structure / content is kind of arising gradually. So everything is prone to change a bit in that sense at this point in time.

File: docs/docs/learn-by-example/1-writing-an-initial-program/README.md (hunk 5 above)
Comment (JorgeCastilloPrz) [reply docs/docs/learn-by-example/1-writing-an-initial-program/README.md]: Great, thanks! :+1: I truly never know how to write this one properly 😅 I'll also ask @MaureenElsberry or @jkmcclellan for a lang review once we complete the tech review.

**Review by rachelcarmena — APPROVED** [review]
Congrats for this awesome new section!! :tada: It will be super useful :raised_hands:

**Comment by JorgeCastilloPrz** [comment]
Thanks @rachelcarmena! 😄 

@MaureenElsberry @jkmcclellan could we get a lang review on these three posts before merging? I want to make sure lang and grammar is fine 🙏

**Comment by jkmcclellan** [comment]
@JorgeCastilloPrz Can I get edit permissions? I don't currently have them for arrow-site.

**Comment by JorgeCastilloPrz** [comment]
Absolutely! I've just granted you and @MaureenElsberry write permissions. Let me know if there's any other issue @jkmcclellan 🙏

**Pull request merged**